src/daemon/daemon_rpc.cpp

```diff
@@ -92,143 +100,219 @@ grpc::Status emit_signal_and_wait_for_result(OperationSignal operation_signal)
-        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_umount, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::version(grpc::ServerContext* context, const VersionRequest* request,
                                     grpc::ServerWriter<VersionReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_version, this, request, response, std::placeholders::_1), client_cert_from(context));
 }
 
 grpc::Status mp::DaemonRpc::ping(grpc::ServerContext* context, const PingRequest* request, PingReply* response)
 {
+    if (connection_type == mp::RpcConnectionType::ssl)
+    {
+        auto client_cert = client_cert_from(context);
+
+        if (!client_cert.empty() && client_cert_store->verify_cert(client_cert))
+        {
+            return grpc::Status::OK;
+        }
+
+        return grpc::Status{grpc::StatusCode::UNAUTHENTICATED, ""};
+    }
+
     return grpc::Status::OK;
 }
 
 grpc::Status mp::DaemonRpc::get(grpc::ServerContext* context, const GetRequest* request,
                                 grpc::ServerWriter<GetReply>* response)
 {
-    return emit_signal_and_wait_for_result(
-        std::bind(&DaemonRpc::on_get, this, request, response, std::placeholders::_1));
+    return verify_client_and_dispatch_operation(
+        std::bind(&DaemonRpc::on_get, this, request, response, std::placeholders::_1), client_cert_from(context));
+}
+
+grpc::Status mp::DaemonRpc::authenticate(grpc::ServerContext* context, const AuthenticateRequest* request,
+                                         grpc::ServerWriter<AuthenticateReply>* response)
+{
+    auto status = emit_signal_and_wait_for_result(
+        std::bind(&DaemonRpc::on_authenticate, this, request, response, std::placeholders::_1));
+
+    if (status.ok() && connection_type == mp::RpcConnectionType::ssl)
+    {
+        const auto store_was_empty{client_cert_store->is_store_empty()};
+
+        client_cert_store->add_cert(client_cert_from(context));
+
+        if (store_was_empty && server_socket_type == mp::ServerSocketType::unix)
+        {
+            MP_PLATFORM.set_server_permissions(server_address, false);
+        }
+    }
+
+    return status;
+}
+
+template <typename OperationSignal>
+grpc::Status mp::DaemonRpc::verify_client_and_dispatch_operation(OperationSignal signal, const std::string& client_cert)
+{
+    if (connection_type == mp::RpcConnectionType::ssl)
+    {
+        if (server_socket_type == mp::ServerSocketType::unix && client_cert_store->is_store_empty())
+        {
+            client_cert_store->add_cert(client_cert);
+            MP_PLATFORM.set_server_permissions(server_address, false);
```

Comment:
This is what I mentioned before that feels "all over the place". The policy on whether the socket is restricted or not is sprinkled in too many places I think - asking for trouble.

Reply:
Well, the problem is that the cert store is platform independent but the type of socket is partially platform dependent.  What I mean by partially is that the Unix socket type is only on Linux and Mac, but the TCP socket type can be supported on all platforms.  So really, the problem is trying to take into account what the socket type is and if the client connecting can have its cert automatically imported depending on what the socket type is and the user's permission on this socket.  It is sprinkled because of the very nature of the problems we are trying to solve.

Reply:
Sure, but we can come up with a way to containerize it. I'm thinking (barely) something on `DaemonRpc` that you can call whenever needed - and then that's what's looking at the server address, cert store and the connection type.

There should only be one method in which we call `MP_PLATFORM.set_server_permission()`. It's not expensive, we only call it on startup and then whenever the conditions (maybe) change.

Reply:
Ok, I did struggle with this as it really is tied to the socket type and nothing else.  One type of socket is platform specific to 2 platforms and the other isn't. But I get what you are saying and I will give it some thought about how this can be improved though.

Reply:
Ok, I cleaned this up some.  Let me know what you think.

Reply:
As discussed, this can be made more simple by requiring client/server connections to **always** be secure and assume that it is.  This will require some changes in tests to use a secure connection.

Reply:
I've committed the changes needed to *always* use a secure connection.  I'm going to close this discussion since this code has changed quite a bit.